### PR TITLE
Pods 3847

### PR DESCRIPTION
**Pull request by vinnicombe**

### CODE REVIEW

File: app/controllers/DataRetrievals.scala [hunk 1]
```diff
@@ -46,6 +45,14 @@ object DataRetrievals {
     }
   }
 
+  def retrievePSTRAndChargeADetails(block: (String, ChargeDetails) => Future[Result])
```
Comment (grantrobson) [inline app/controllers/DataRetrievals.scala]: As this is specific to charge A and is only used on charge A's CYA controller it perhaps is not necessary to have this method on this trait, instead I'd just put it on the CYA controller and make it private.

File: app/controllers/DataRetrievals.scala [hunk 2]
```diff
@@ -62,7 +69,16 @@ object DataRetrievals {
     }
   }
 
-  def retrieveSchemeAndSponsoringEmployer(block: (String,String) => Future[Result])(implicit request: DataRequest[AnyContent]): Future[Result] = {
+  def retrieveSchemeAndChargeADetails(block: (String, ChargeDetails) => Future[Result])
```
Comment (grantrobson) [inline app/controllers/DataRetrievals.scala]: As above

File: app/controllers/chargeA/ChargeDetailsController.scala [hunk 3]
```diff
@@ -52,6 +52,13 @@ class ChargeDetailsController @Inject()(override val messagesApi: MessagesApi,
   def form()(implicit messages: Messages): Form[ChargeDetails] =
     formProvider()
 
+  def viewModel(mode: Mode, srn: String, schemeName: String): GenericViewModel =
```
Comment (grantrobson) [inline app/controllers/chargeA/ChargeDetailsController.scala]: Make this private

File: app/forms/mappings/Formatters.scala [hunk 4]
```diff
@@ -186,11 +247,16 @@ trait Formatters extends Transforms with Constraints {
...
-        val total = itemsToTotal.foldLeft[BigDecimal](BigDecimal(0)) { (acc, next) =>
-          Try(BigDecimal(data.getOrElse(next, "0"))).toOption.fold(BigDecimal(0))(_ + acc)
-        }
+        val total =
+          itemsToTotal.foldLeft[BigDecimal](BigDecimal(0)) { (acc, next) =>
+            data.get(next) match {
+              case Some(str) =>
+                Try(BigDecimal(str)).getOrElse(BigDecimal(0.00)) + acc
+              case None =>
+                BigDecimal(0.00) + acc
```
Comment (grantrobson) [inline app/forms/mappings/Formatters.scala]: BigDecimal(0.00) is redundant here - you can just remove it.

File: test/forms/chargeA/ChargeDetailsFormProviderSpec.scala [hunk 5]
```diff
@@ -30,88 +29,131 @@ class ChargeDetailsFormProviderSpec extends DateBehaviours with BigDecimalFieldB
...
   private val totalAmtOfTaxDueAtLowerRateKey = "totalAmtOfTaxDueAtLowerRate"
   private val totalAmtOfTaxDueAtHigherRateKey = "totalAmtOfTaxDueAtHigherRate"
-  private val totalAmtKey = "totalAmount"
-
   private val messageKeyNumberOfMembersKey = "chargeA.numberOfMembers"
   private val messageKeyAmountTaxDueLowerRateKey = "chargeA.totalAmtOfTaxDueAtLowerRate"
   private val messageKeyAmountTaxDueHigherRateKey = "chargeA.totalAmtOfTaxDueAtHigherRate"
 
-  private val decimalFormat = new DecimalFormat("0.00")
+  def chargeADetails(
```
Comment (grantrobson) [inline test/forms/chargeA/ChargeDetailsFormProviderSpec.scala]: Make private

File: app/controllers/DataRetrievals.scala (hunk 1 above)
Comment (vinnicombe) [reply app/controllers/DataRetrievals.scala]: Done

File: app/controllers/DataRetrievals.scala (hunk 2 above)
Comment (vinnicombe) [reply app/controllers/DataRetrievals.scala]: Done

File: app/controllers/chargeA/ChargeDetailsController.scala (hunk 3 above)
Comment (vinnicombe) [reply app/controllers/chargeA/ChargeDetailsController.scala]: Done

File: app/forms/mappings/Formatters.scala (hunk 4 above)
Comment (vinnicombe) [reply app/forms/mappings/Formatters.scala]: Done

File: test/forms/chargeA/ChargeDetailsFormProviderSpec.scala (hunk 5 above)
Comment (vinnicombe) [reply test/forms/chargeA/ChargeDetailsFormProviderSpec.scala]: Done

**Pull request merged**